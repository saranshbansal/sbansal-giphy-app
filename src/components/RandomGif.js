import React, { Component } from "react";
import Gif from "./Gif";
import Loader from "./library/Loader";
import { fetchRandom } from "../resources/api";

class RandomGif extends Component {
  constructor(props) {
    super(props);

    this.refreshResults = this.refreshResults.bind(this);
  }

  state = {
    loading: false,
    gif: null
  };

  componentDidMount() {
    this.refreshResults();
  }

  async refreshResults() {
    let gifContent = null;

    try {
      this.setState({
        loading: true
      });

      const result = await fetchRandom();

      if (!!result && result.data) {
        gifContent = {
          id: result.data.id,
          imageUrl: result.data.images.downsized_large.url,
          staticImageUrl: result.data.images.downsized_still.url,
          title: result.data.title
        };
      }
    } catch (e) {
      console.error(e);
    } finally {
      this.setState({
        loading: false,
        gif: gifContent
      });
    }
  }

  render() {
    const { gif, loading } = this.state;

    const showResults = gif && !loading;

    return (
      <div className="random">
        {!showResults && <Loader />}

        <button className="btn" onClick={this.refreshResults} aria-label="Randomise">
          RANDOMISE
        </button>

        {showResults && <Gif data={gif} />}
      </div>
    );
  }
}

export default RandomGif;
