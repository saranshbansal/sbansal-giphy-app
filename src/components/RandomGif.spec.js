import React from "react";
import { shallow, mount } from "enzyme";
import RandomGif from "./RandomGif";

describe("Generate Random Gif", () => {
  it("should render correctly.", () => {
    const wrapper = shallow(<RandomGif />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should load a random GIF on calling handleSearch fn.", async () => {
    const wrapper = mount(<RandomGif />);
    const instance = wrapper.instance();

    expect(wrapper.state("gif")).toBe(null);

    await instance.refreshResults();

    expect(wrapper.state("gif")).not.toBe(null);
  });
});
