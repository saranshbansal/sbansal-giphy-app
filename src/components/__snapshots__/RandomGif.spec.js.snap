// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Random Gif should render correctly. 1`] = `
<div
  className="random"
>
  <Loader />
  <button
    className="btn"
    onClick={[Function]}
  >
    RANDOMISE
  </button>
</div>
`;
